cdap-data-fabric/src/main/java/co/cask/cdap/spi/data/common/MetricStructuredTable.java

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.spi.data.common;
+
+import co.cask.cdap.api.dataset.lib.CloseableIterator;
+import co.cask.cdap.api.metrics.MetricsCollector;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.spi.data.InvalidFieldException;
+import co.cask.cdap.spi.data.StructuredRow;
+import co.cask.cdap.spi.data.StructuredTable;
+import co.cask.cdap.spi.data.table.StructuredTableId;
+import co.cask.cdap.spi.data.table.field.Field;
+import co.cask.cdap.spi.data.table.field.Range;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Optional;
+
+/**
+ * Structured table that takes a delegation and emit metrics on each operation.
+ */
+public class MetricStructuredTable implements StructuredTable {
+  private final StructuredTable structuredTable;
+  private final MetricsCollector metricsCollector;
+  private final String metricPrefix;
+  private final boolean emitTimeMetrics;
+
+  public MetricStructuredTable(StructuredTableId tableId,
+                               StructuredTable structuredTable, MetricsCollector metricsCollector,
+                               boolean emitTimeMetrics) {
+    this.structuredTable = structuredTable;
+    this.metricsCollector = metricsCollector;
+    this.metricPrefix = Constants.Metrics.StructuredTable.METRICS_PREFIX + tableId.getName() + ".";
+    this.emitTimeMetrics = emitTimeMetrics;
+  }
+
+  @Override
+  public void upsert(Collection<Field<?>> fields) throws InvalidFieldException, IOException {
+    try {
+      if (!emitTimeMetrics) {
+        structuredTable.upsert(fields);
+      } else {
+        long curTime = System.currentTimeMillis();
```

Comment:
Use `System.nanoTime` to measure duration rather than `currentTimeMillis`.

Reply:
Fixed